### PR TITLE
[HOP][print]Add grad printing with flag

### PR DESCRIPTION
**Pull request by fxdawnn**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #175224
* #175552

## Summary                                                                                
                                                                                         
  Add a print_backward flag to the torch._higher_order_ops.print HOP. When
  print_backward=True, the HOP returns tensor arguments (identity pass-through) so they
  stay in the autograd graph, and prints gradients during backward with a [backward]
  prefix. This works under torch.compile with fullgraph=True (no graph break).

 ## Usage
```
  import torch
  import torch._higher_order_ops

  // Forward-only print (existing behavior, returns None)
  torch._higher_order_ops.print("x: {}", x)

  // Forward + backward print (returns tensor args, prints grads in backward)
  x = torch._higher_order_ops.print("x: {}", x, print_backward=True)
  // Forward output:  "x: tensor([1., 2., 3.])"
  // Backward output: "[backward] grad tensor([2., 2., 2.])"

  // In a training loop:
  class Model(torch.nn.Module):
      def forward(self, x):
          h = self.linear1(x)
          h = torch._higher_order_ops.print("after linear1: {}", h, print_backward=True)
          h = self.linear2(h)
          h = torch._higher_order_ops.print("after linear2: {}", h, print_backward=True)
          return h

  model = Model()
  opt_model = torch.compile(model, fullgraph=True)
  out = opt_model(x)
  loss = F.mse_loss(out, target)
  loss.backward()
  // Prints activations in forward, gradients in backward — no graph break.
```
  ## Test plan

  - TestHopPrintBackward — 8 tests covering basic backward, gradient passthrough, return
  semantics, compiled (eager/aot_eager with fullgraph=True), and a realistic train step

Screenshot 2026-02-17 at 7 22 42 PM

### CODE REVIEW

File: torch/_higher_order_ops/print.py [hunk 1]
```diff
@@ -26,17 +40,61 @@ class Print(HigherOrderOperator):
        torch._higher_order_ops.print("moo {} {y}", 1, y=2)
        Output: "moo 1 2"
 
+    4. Gradient printing during backward (print_backward=True):
+       x = torch._higher_order_ops.print("x: {}", x, print_backward=True)
```
Comment (yushangdi) [inline torch/_higher_order_ops/print.py:54]: does this only work if you do `x = ...`? what if you're printing multiple tensors?

Comment (fxdawnn) [reply torch/_higher_order_ops/print.py:54]: we might have to do the same thing for as many as tensor we want. e.g., do ``y = torch._higher_order_ops.print("y: {}", y, print_backward=True)`` for the next one...

I can't think of a better way to pass tensor to autograd besides this setup...

Comment (angelayi) [reply torch/_higher_order_ops/print.py:54]: +1 I'm not sure about the design of this. also I don't think it's natural for people to have to use the tensor returned from `print`?

Comment (yushangdi) [reply torch/_higher_order_ops/print.py:54]: my proposal let's keep this print HOP simple and straightforward, and use it only for forward. If people want to print in backward, they can use the more complicated version here: https://github.com/pytorch/pytorch/pull/174636  (PR not ready for review yet as I got a bunch of merge conflict with main and I'm resolving them)

I got the backward hook to work for compile full graph, it can print gradients of intermediate tensors (with the caveat that it cannot print the gradient of the input tensors)

Comment (fxdawnn) [reply torch/_higher_order_ops/print.py:54]: This one shangdi shared looks more like a hash print! 

I observed that the solution of adding hook might lead to graph break but totally agree that could be the move. This might be the cleanest solution among all to get the grad printed. But I can also turn it into stale for now.

Comment (yushangdi) [reply torch/_higher_order_ops/print.py:54]: @fxdawnn do you have an example for the graph break?

Comment (fxdawnn) [reply torch/_higher_order_ops/print.py:54]: no! But i don't think the hash_print and printing grad here with a flag is conflicting. They're complimentary to each other
